test/extensions/filters/udp/udp_proxy/session_filters/http_capsule/http_capsule_filter_test.cc

```diff
@@ -0,0 +1,213 @@
+namespace SessionFilters {
+namespace HttpCapsule {
+namespace {
+
+class HttpCapsuleFilterTest : public testing::Test {
+public:
+  void setup() {
+    filter_ = std::make_unique<HttpCapsuleFilter>(server_context_.timeSource());
+    filter_->initializeReadFilterCallbacks(read_callbacks_);
+    filter_->initializeWriteFilterCallbacks(write_callbacks_);
+  }
+
+  NiceMock<Server::Configuration::MockFactoryContext> server_context_;
+  std::unique_ptr<HttpCapsuleFilter> filter_;
+  NiceMock<MockReadFilterCallbacks> read_callbacks_;
+  NiceMock<MockWriteFilterCallbacks> write_callbacks_;
+  NiceMock<StreamInfo::MockStreamInfo> stream_info_;
+};
+
+TEST_F(HttpCapsuleFilterTest, ContinueOnNewSession) {
+  setup();
+  EXPECT_EQ(ReadFilterStatus::Continue, filter_->onNewSession());
+}
+
+TEST_F(HttpCapsuleFilterTest, EncapsulateEmptyDatagram) {
+  setup();
+
+  Network::UdpRecvData datagram;
+  datagram.buffer_ = std::make_unique<Buffer::OwnedImpl>();
+  EXPECT_EQ(ReadFilterStatus::Continue, filter_->onData(datagram));
+
+  const std::string expected_data = absl::HexStringToBytes("00" // DATAGRAM Capsule Type
+                                                           "01" // Context ID length
+                                                           "00" // Context ID
+  );
+
+  EXPECT_EQ(expected_data, datagram.buffer_->toString());
+}
+
+TEST_F(HttpCapsuleFilterTest, EncapsulateDatagram) {
+  setup();
+
+  const std::string payload = "payload";
+  Network::UdpRecvData datagram;
+  datagram.buffer_ = std::make_unique<Buffer::OwnedImpl>();
+  datagram.buffer_->add(payload);
+  EXPECT_EQ(ReadFilterStatus::Continue, filter_->onData(datagram));
+
+  const std::string expected_data =
+      absl::HexStringToBytes("00" // DATAGRAM Capsule Type
+                             "08" // Capsule Length = length(payload) + 1
+                             "00" // Context ID
+                             ) +
+      payload;
+
+  EXPECT_EQ(expected_data, datagram.buffer_->toString());
+}
+
+TEST_F(HttpCapsuleFilterTest, InvalidCapsule) {
+  setup();
+
+  const std::string invalid_context_id_fragment =
+      absl::HexStringToBytes("00" // DATAGRAM Capsule Type
+                             "01" // Capsule Length
+                             "c0" // Context ID (Invalid VarInt62)
+      );
+
+  Network::UdpRecvData datagram;
+  datagram.buffer_ = std::make_unique<Buffer::OwnedImpl>();
+  datagram.buffer_->add(invalid_context_id_fragment);
+  EXPECT_CALL(write_callbacks_, injectDatagramToFilterChain(_)).Times(0);
+  EXPECT_EQ(WriteFilterStatus::StopIteration, filter_->onWrite(datagram));
+  EXPECT_EQ(0, datagram.buffer_->length());
+}
+
+TEST_F(HttpCapsuleFilterTest, IncompatibleCapsule) {
+  setup();
+
+  const std::string invalid_context_id_fragment =
+      absl::HexStringToBytes("01" // Capsule Type is not DATAGRAM
+                             "01" // Capsule Length
+                             "00" // Context ID (Invalid VarInt62)
+      );
```

Comment:
This test isn't quite right. When the capsule type is 1, the payload doesn't start with a context ID. For that matter, `01` is [already used for something else](https://www.iana.org/assignments/masque/masque.xhtml). Additionally, this isn't invalid - it's just unknown. To avoid future issues, I'd recommend using a reserved value:

```suggestion
  const std::string unexpected_capsule_fragment =
      absl::HexStringToBytes("17" // Capsule Type is not DATAGRAM
                             "01" // Capsule Length
                             "00" // Unknown capsule payload
      );
```